src/spot_skills/src/spot_skills_py/spot/spot_manager.py

```
@@ -139,56 +206,135 @@ def log_lease_info(self) -> None:
+        try:
+            # Try returning the specific lease that we hold
+            with contextlib.suppress(Exception):
+                lease = self._lease_wallet.get_lease(self._resource)
+                self._lease_client.return_lease(lease)
+
+            if self._lease_keeper:
+                self._lease_keeper.shutdown()  # Blocks until complete
+                self._lease_keeper = None
+
+        finally:
+            self._control_status = self._compute_status()
+
+    @property
+    def has_control(self) -> bool:
+        """Check whether the SpotManager has control of Spot right now."""
+        self._control_status = self._compute_status()
+        return self._control_status.ok
+
+    def _compute_status(self) -> SpotControlStatus:
+        """Compute the current control status of the SpotManager."""
+        powered_on = False
+        try:
+            powered_on = self._robot.is_powered_on()
+        except Exception:
+            powered_on = False
+
+        lease_owner = None
+        lease_epoch = None
+        has_lease = False
+
+        try:  # Prefer the wallet as the ground truth for leases
+            state = self._lease_wallet.get_lease_state(self._resource)
+
+            status_enum = getattr(state, "lease_status", None)
+            owner = getattr(state, "lease_owner", None)
+            lease_owner = getattr(state, "client_name", None) or (
+                getattr(owner, "client_name", None) if owner else None
+            )
+
+            lease_msg = getattr(state, "lease_current", None) or getattr(state, "lease", None)
+            lease_epoch = getattr(lease_msg, "epoch", None)
+
+            has_lease = (status_enum is not None) and (
+                int(status_enum) == int(type(state).Status.SELF_OWNER)
+            )
+            # Double-check that the KeepAlive thread is up
+            has_lease = has_lease and self._holds_live_lease()
+
+        except Exception:
+            has_lease = self._holds_live_lease()
```

At least print ignored exceptions, in case they're silently always happening.